dsp/src/lockin.rs

```diff
@@ -0,0 +1,534 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `phase_offset` - Phase offset (in radians) applied to the
+    /// demodulation signal.
+    /// * `sample_period` - ADC sampling period in terms of the
+    /// internal clock period.
+    /// * `harmonic` - Integer scaling factor used to adjust the
+    /// demodulation frequency. E.g., 2 would demodulate with the
+    /// first harmonic.
+    /// * `iir` - IIR biquad filter. Two identical copies of this IIR
+    /// filter are used: one for the in-phase signal and the other for
+    /// the quadrature signal.
+    ///
+    /// # Returns
+    ///
+    /// New `Lockin` instance.
+    pub fn new(
+        phase_offset: f32,
+        sample_period: u32,
+        harmonic: u32,
+        iir: IIR,
+    ) -> Self {
+        Lockin {
+            phase_offset: phase_offset,
+            sample_period: sample_period,
+            harmonic: harmonic,
+            timestamps: [None, None],
+            iir: [iir, iir],
+            iirstate: [[0.; 5]; 2],
+        }
+    }
+
+    /// Demodulate an input signal with in-phase and quadrature
+    /// reference signals.
+    ///
+    /// # Arguments
+    ///
+    /// * `adc_samples` - One batch of ADC samples.
+    /// * `timestamps` - Counter values corresponding to the edges of
+    /// an external reference signal. The counter is incremented by a
+    /// fast internal clock.
+    /// * `valid_timestamps` - The number of valid timestamps in
+    /// `timestamps`. Only `&timestamps[..valid_timestamps]` are used;
+    /// every other value in the `timestamps` array is ignored.
+    ///
+    /// # Returns
+    ///
+    /// The demodulated in-phase and quadrature signals as an
+    /// `Option`. When there are an insufficient number of timestamps
+    /// to perform processing, `None` is returned.
+    ///
+    /// # Assumptions
+    ///
+    /// `demodulate` expects that the timestamp counter value is equal
+    /// to 0 when the ADC samples its first input in a batch. This can
+    /// be achieved by configuring the timestamp counter to overflow
+    /// at the end of the ADC batch sampling period.
+    pub fn demodulate(
+        &mut self,
+        adc_samples: [i16; ADC_SAMPLE_BUFFER_SIZE],
+        timestamps: [u16; TIMESTAMP_BUFFER_SIZE],
+        valid_timestamps: u16,
+    ) -> Option<([f32; ADC_SAMPLE_BUFFER_SIZE], [f32; ADC_SAMPLE_BUFFER_SIZE])>
+    {
+        // update old timestamps for new ADC batch
+        let sample_period = self.sample_period as i32;
+        self.timestamps.iter_mut().for_each(|t| match *t {
+            Some(i) => {
+                *t = Some(i - ADC_SAMPLE_BUFFER_SIZE as i32 * sample_period);
+            }
+            None => (),
+        });
+
+        // record new timestamps
+        timestamps
+            .iter()
+            .take(valid_timestamps as usize)
+            .rev()
+            .take(2)
+            .rev()
+            .for_each(|t| self.timestamps.push(Some(*t as i32)));
```

Comment:
Won't this infinitely grow `self.timestamps`? I'm not seeing a truncation step anywhere.

Reply:
I guess I'm not following what `push` does here? Isn't `self.timestamps` just an array? How can we "push" to it?

Reply:
I found the trait below - would recommend moving the definition above so that others aren't as confused as I was